store/postgres/src/metadata.rs

```diff
@@ -409,3 +425,270 @@ pub fn deployment_state_from_id(
         }
     }
 }
+
+/// Delete all assignments for deployments that are neither the current nor the
+/// pending version of a subgraph and return the deployment id's
+fn remove_unused_assignments(conn: &PgConnection) -> Result<Vec<EntityChange>, StoreError> {
+    const QUERY: &str = "
+    delete from subgraphs.subgraph_deployment_assignment a
+    where not exists (select 1
+                        from subgraphs.subgraph s, subgraphs.subgraph_version v
+                       where v.id in (s.current_version, s.pending_version)
+                         and v.deployment = a.id)
+    returning a.id
+    ";
+    #[derive(QueryableByName)]
+    struct Removed {
+        #[sql_type = "Text"]
+        id: String,
+    }
+
+    Ok(diesel::sql_query(QUERY)
+        .load::<Removed>(conn)?
+        .into_iter()
+        .map(|r| {
+            MetadataOperation::Remove {
+                entity: SubgraphDeploymentAssignmentEntity::TYPENAME,
+                id: r.id,
+            }
+            .into()
+        })
+        .filter_map(|e| e)
+        .collect::<Vec<_>>())
+}
+
+/// Mark the deployment `id` as synced, and promote it to the current version
+/// everywhere where it was the pending version so far, and remove any
+/// assignments that are not needed any longer as a result. Return the changes
+/// that were made to assignments in the process
+pub fn deployment_synced(
+    conn: &PgConnection,
+    id: &SubgraphDeploymentId,
+) -> Result<Vec<EntityChange>, StoreError> {
+    use subgraph as s;
+    use subgraph_deployment as d;
+    use subgraph_version as v;
+
+    // Subgraphs where we need to promote the version
+    let pending_subgraph_versions: Vec<(String, String)> = s::table
+        .inner_join(v::table.on(s::pending_version.eq(v::id.nullable())))
+        .filter(v::deployment.eq(id.as_str()))
+        .select((s::id, v::id))
+        .for_update()
+        .load(conn)?;
+
+    // Switch the pending version to the current version
+    for (subgraph, version) in pending_subgraph_versions {
+        update(s::table.filter(s::id.eq(&subgraph)))
+            .set((
+                s::current_version.eq(&version),
+                s::pending_version.eq::<Option<&str>>(None),
+            ))
+            .execute(conn)?;
+    }
+
+    let changes = remove_unused_assignments(conn)?;
```

Comment:
Nice that we are doing this automatically now!

Reply:
This only gets rid of the `SubgraphDeploymentAssignment`, but not the deployments related to them - they still hang around. The change from how we did things before is that before we carefully tried to determine which assignments are unneeded based on how deployments got shuffled around. `remove_unused_assignments` gets rid of any assignment that is not related to any current or pending version, regardless of other operations we were doing.

Reply:
Nice I always feared that piece of the code.